### PR TITLE
Updated door system

### PR DESCRIPTION
**Pull request by mcNuggets1**
+ 1 new hook

* Redid unownalldoors command and removed unnecessary ents.GetAll call and made it work with hooks used by the regular door selling.
(note: this function was disgusting regarding performance)

* Removed unnecessary ents.GetAll call in keysUnOwnAll. It's only used to unlock doors which the other loop can also do.

* Use hook.Call instead of hook.Run because they are gamemode hooks.

* Code cleanup

* Suppress tax message when not unowning anything

Hope you like it <333

### CODE REVIEW

**Comment by mcNuggets1** [comment]
Should be working for now.
Just tested in listen server.

**Comment by mcNuggets1** [comment]
Oh, yes.
Sorry, I just woke up there.

I will revert the hook changes.

**Comment by Bo98** [comment]
Seems fine. The hooks and the added parameter should be documented with stubs at some point.

**Comment by mcNuggets1** [comment]
I never did that.
So I don't know how it would look.

File: gamemode/modules/doorsystem/sv_doors.lua [hunk 1]
```diff
@@ -84,13 +87,16 @@ function meta:keysOwn(ply)
...
 
-function meta:keysUnOwn(ply)
+function meta:keysUnOwn(ply, tax)
     if not ply then
         ply = self:getDoorOwner()
 
         if not IsValid(ply) then return end
     end
 
+    local isAllowed = hook.Call("canUnOwn" .. (self:IsVehicle() and "Vehicle" or "Door"), nil, ply, self, tax)
```
[the remark this answers is not in the document]
Comment (mcNuggets1) [reply gamemode/modules/doorsystem/sv_doors.lua]: I'll think of something.
Just give me some time.

**Comment by FPtje** [comment]
The `keysUnOwn` function still has a tax parameter and calls a hook to fail silently if the hook says no. Is that intentional?

**Comment by mcNuggets1** [comment]
It wass the only efficient way I could find.

You have to call keysUnOwn with a boolean as second argument to call the canTaxRemove hook which determines if the door is getting unowned or not.
It only affects the tax doors and nothing else.

**Comment by FPtje** [comment]
Wait, do you want specific doors not to be sold through the tax mechanism? As opposed to the tax system not selling anything?

**Comment by mcNuggets1** [comment]
Yes.
That's my intention.

**Comment by FPtje** [comment]
But players still pay tax over them? I'm confused about your use case. Can you tell me why you want this hook? What feature are you building for your own server? Isn't it better for there to be some way to exclude doors from the tax system entirely? As opposed to just the "selling because you don't have the money" part.

**Comment by mcNuggets1** [comment]
Doors/Vehicles are only sold, if you don't have enough money to pay them, I want this feature to prevent vehicles from being sold, because I don't use DarkRP's regular vehicle spawning.

Preventing everything to be not sold is not my intention, but to prevent only vehicles from being sold.

**Comment by FPtje** [comment]
Should people pay door tax on vehicles?

**Comment by mcNuggets1** [comment]
They are doing it for some reason, I don't have anything against it, because it makes sense paying for vehicles being on the map.

On second thought can someone prevent 1 vehicle from being taxed?

**Comment by FPtje** [comment]
I'm thinking of an alternative solution that has modders decide which entities get taxed. It's more generic, seems to solve your use case and doesn't hurt the cohesion of the unown function.

**Comment by mcNuggets1** [comment]
You can already override taxes with the canProperyTax hook.

**Comment by mcNuggets1** [comment]
I added the hook in the function keysUnOwnAll instead, which makes more sense, because it actually checks for the door count to suppress the taxes message.

**Comment by FPtje** [comment]
Still, you want to make `keysUnOwnAll` not to unown all.

**Comment by Bo98** [comment]
I suppose the problem extends to a lot of other player functions in DarkRP. Though that's likely something that can't be changed now.

**Comment by FPtje** [comment]
Do you mean that many player functions have the same design error? It's possible, and indeed hard to change if true. Still, I'd like to not make it worse.

**Comment by mcNuggets1** [comment]
Only of you use the functiom with the first argument being true, then the hook should be called. I may have forgotten to implement it that way.

It shouldnt affect the normal usage however

**Comment by FPtje** [comment]
Whether it does it with the argument or not doesn't matter. It's a hack from the tax system to make it do something that contradicts the name of the function.

**Comment by mcNuggets1** [comment]
Nah, I integrated it, there is no way without using an argument on the function for it to not work.

**Comment by FPtje** [comment]
> there is no way without using an argument on the function for it to not work.
Irrelevant to my point.

> there is a way using an argument on the function to make it not work.

This is my point

**Comment by mcNuggets1** [comment]
I could make a local function instead for the tax system, is that alright?

**Comment by FPtje** [comment]
Like I said, I'd rather have a system that allows developers to exclude (or include?) things from the tax system entirely. A local function would be better though.

**Comment by Bo98** [comment]
>  Still, I'd like to not make it worse.

Indeed.

It looks like here what is desired is a filter. `Player:keysUnOwnAll` shouldn't filter things - instead it should be separate and then call `Player:keysUnOwn`.

In addition, the idea of having it be there for one specific system that happens to call the function doesn't make sense to me. Why should `Player:keysUnOwnAll` have logic for the tax system specifically? It should be completely independent of whatever system is calling it.

**Comment by mcNuggets1** [comment]
It says this:

if ply:getDarkRPVar("money") < (GAMEMODE.Config.startingmoney * 2) then return false end

This means, the taxes go never below this point, meaning no player can loose their property, which isn't entirely true.

**Comment by Bo98** [comment]
Wallet tax and property tax are different things. canTax controls wallet tax only.

**Comment by mcNuggets1** [comment]
Oh, I totally forgot about this.
Thanks for clarifying.

**Comment by Bo98** [comment]
The wallet tax system should not be in the doorsystem module tbh. It doesn't make much sense there.

**Comment by mcNuggets1** [comment]
Yeah, It shouldn't, I already missinterpretated that.

However, we are using a local function now.

File: gamemode/modules/doorsystem/sv_doors.lua [hunk 2]
```diff
@@ -397,18 +418,34 @@ DarkRP.defineChatCommand("toggleown", OwnDoor)
...
+
+    if ply.Ownedz then
+        for k, v in pairs(ply.Ownedz) do
+            if not v:isKeysOwnable() then ply.Ownedz[k] = nil continue end
+            local bAllowed, strReason = hook.Call("playerSell" .. (v:IsVehicle() and "Vehicle" or "Door"), GAMEMODE, ply, v)
+
+            if bAllowed == false then
+                if strReason and strReason ~= "" then
+                    DarkRP.notify(ply, 1, 4, strReason)
+                end
```
[the remark this answers is not in the document]
Comment (mcNuggets1) [reply gamemode/modules/doorsystem/sv_doors.lua]: Isn't this in default DarkRP already?

**Comment by mcNuggets1** [comment]
I updated it.

I'm doing single pull requests, because I didnt know how to do multiple, sorry.

**Comment by FPtje** [comment]
Sorry, I wasn't clear. I'm already working on it. I'm doing quite a few changes. If you don't mind too much, I'll rewrite the commits.

**Comment by mcNuggets1** [comment]
Just do as you like.

File: gamemode/modules/doorsystem/sv_doors.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (FPtje) [reply gamemode/modules/doorsystem/sv_doors.lua]: No, in previous versions DarkRP didn't call the hook.

**Comment by FPtje** [comment]
Check the above commits. I've written it all in one go, but split it up in three distinct commits using GitSavvy.

**Pull request closed without merging**